spring-rabbit-stream/src/main/java/org/springframework/rabbit/stream/producer/RabbitStreamTemplate.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.rabbit.stream.producer;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessagePostProcessor;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.amqp.support.converter.SimpleMessageConverter;
+import org.springframework.beans.factory.BeanNameAware;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.lang.Nullable;
+import org.springframework.rabbit.stream.support.StreamMessageProperties;
+import org.springframework.rabbit.stream.support.converter.DefaultStreamMessageConverter;
+import org.springframework.rabbit.stream.support.converter.StreamMessageConverter;
+import org.springframework.util.Assert;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+import com.rabbitmq.stream.ConfirmationHandler;
+import com.rabbitmq.stream.Constants;
+import com.rabbitmq.stream.Environment;
+import com.rabbitmq.stream.MessageBuilder;
+import com.rabbitmq.stream.Producer;
+import com.rabbitmq.stream.ProducerBuilder;
+
+/**
+ * @author Gary Russell
+ * @since 2.8
+ *
+ */
+public class RabbitStreamTemplate implements RabbitStreamOperations, BeanNameAware, DisposableBean {
+
+	private final Environment environment;
+
+	private final String streamName;
+
+	private MessageConverter messageConverter = new SimpleMessageConverter();
+
+	private StreamMessageConverter streamConverter = new DefaultStreamMessageConverter();
+
+	private Producer producer;
+
+	private String beanName;
+
+	private ProducerCustomizer producerCustomizer = (name, builder) -> { };
+
+	/**
+	 * Construct an instance with the provided {@link Environment}.
+	 * @param environment the environment.
+	 * @param streamName the stream name.
+	 */
+	public RabbitStreamTemplate(Environment environment, String streamName) {
+		this.environment = environment;
+		this.streamName = streamName;
+	}
+
+
+	private synchronized Producer createOrGetProducer() {
```

Comment:
Why can't we do this in the `afterPropertiesSet()`?
I can guess because this template might be used outside of Spring management, but then that `DisposableBean` contract is still not going to be satisfied...

Reply:
1. Because it will cause a connection to the broker and `InitializingBean` is too soon.
2. I added `reset()` for programmatic - I will change it to `AutoCloseable`.